### PR TITLE
Add ovf tests and more overflow macro definitions ##util

### PR DESCRIPTION
**Pull request by trufae**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**
There are many places in r2 that add, sub and mul overflow and underflows can happen. we want to provide macros to detect such events and act accordingly.

This PR adds some more macros and implements a bunch of tests

* https://www.fefe.de/intof.html
* https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html

```c
// for addition
#include <limits.h>
int a = <something>;
int x = <something>;
if ((x > 0) && (a > INT_MAX - x)) /* `a + x` would overflow */;
if ((x < 0) && (a < INT_MIN - x)) /* `a + x` would underflow */;

// for subtraction
#include <limits.h>
int a = <something>;
int x = <something>;
if ((x < 0) && (a > INT_MAX + x)) /* `a - x` would overflow */;
if ((x > 0) && (a < INT_MIN + x)) /* `a - x` would underflow */;

// for multiplication
#include <limits.h>
int a = <something>;
int x = <something>;
if (a > INT_MAX / x) /* `a * x` would overflow */;
if ((a < INT_MIN / x)) /* `a * x` would underflow */;

// there may be need to check for -1 for two's complement machines
if ((a == -1) && (x == INT_MIN)) /* `a * x` can overflow */
if ((x == -1) && (a == INT_MIN)) /* `a * x` (or `a / x`) can overflow */
```

**Test plan**

unit tests. this is needed for another PR

**Closing issues**

https://github.com/radareorg/radare2/pull/17188#discussion_r449814094

### CODE REVIEW

**Comment by XVilka** [comment]
https://github.com/radareorg/radare2/pull/17216/checks?check_run_id=839529545#step:13:699
```
make[1]: Entering directory '/home/runner/work/radare2/radare2/test/unit'
mkdir -p "bin"
cc test_parse_ctype.c -o bin/test_parse_ctype -I/home/runner/include/libr -g -L/home/runner/lib/x86_64-linux-gnu -lr_core -lr_fs -lr_debug -lr_hash -lr_anal -lr_search -lr_egg -lr_crypto -lr_config -lr_bin -lr_magic -lr_io -lr_asm -lr_parse -lr_syscall -lr_reg -lr_lang -lr_cons -lr_flag -lr_socket -lr_bp -lr_util
In file included from /home/runner/include/libr/r_types.h:207:0,
                 from /home/runner/include/libr/r_parse.h:6,
                 from test_parse_ctype.c:1:
/home/runner/include/libr/r_types_base.h:105:10: fatal error: r_types_overflow.h: No such file or directory
 #include "r_types_overflow.h"
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [bin/test_parse_ctype] Error 1
```

File: libr/include/r_types_overflow.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#ifndef R_TYPES_OVERFLOW_H
+#define R_TYPES_OVERFLOW_H
+
+/* preventive math overflow checks */
+#if !defined(SZT_ADD_OVFCHK)
+#define SZT_ADD_OVFCHK(x,y) ((SIZE_MAX - (x)) < (y))
+#endif
+
+#if 0
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:9]: just use comments instead of `#if 0` if you want to explain something.

File: libr/include/r_types_overflow.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// ADD
+if ((x > 0) && (a > INT_MAX - x)) /* `a + x` would overflow */;
+if ((x < 0) && (a < INT_MIN - x)) /* `a + x` would underflow */;
+#endif
+#define UT64_ADD_OVFCHK(x,y) ((UT64_MAX - (x)) < (y))
+#define ST64_ADD_OVFCHK(a,x) ((((x) > 0) && ((a) > ST64_MAX - (x))) || ((x) <0 &&(a)< ST32_MIN - (x)))
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:15]: check style, in particular near `... &&(a)...`, here and in next lines.

File: libr/include/r_types_overflow.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+#define UT64_MUL_OVFCHK(x,y) ((x) > (UT64_MAX / (y)))
+#define ST64_MUL_OVFCHK(x,y) ((x) > (ST64_MAX / (y)))
+#define UT32_MUL_OVFCHK(x,y) ((x) > (UT32_MAX / (y)))
+#define ST32_MUL_OVFCHK(x,y) ((x) > (ST32_MAX / (y)))
+#define UT16_MUL_OVFCHK(x,y) ((x) > (UT16_MAX / (y)))
+#define ST16_MUL_OVFCHK(x,y) ((x) > (ST16_MAX / (y)))
+#define UT8_MUL_OVFCHK(x,y) ((x) > (UT8_MAX / (y)))
+#define ST8_MUL_OVFCHK(x,y) ((x) > (ST8_MAX / (y)))
+// #define ST8_MUL_OVFCHK(a,x) ((x>0) && (a)>(ST8_MIN / (x)))
+#if 0
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:42]: why putting this comments at the end? I would place it where they are more relevant and remove the `#if 0`.

File: libr/include/r_types_overflow.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+// SUB
+if ((x < 0) && (a > INT_MAX + x)) /* `a - x` would overflow */;
+if ((x > 0) && (a < INT_MIN + x)) /* `a - x` would underflow */;
+// MUL
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:46]: the macros defined do not exactly use this code (they don't check for -1 or INT_MIN, for example, why?

File: test/unit/test_ovf.c [hunk 5]
```diff
@@ -0,0 +1,125 @@
+#include <r_util.h>
+#include "minunit.h"
+
+int test_overflow_add(void) {
+	int a = 250;
+	int b = 32;
+	if (UT8_ADD_OVFCHK(a, b)) {
```
Comment (ret2libc) [inline test/unit/test_ovf.c:7]: these tests look wrong :/

Just do `mu_assert_false (UT8_ADD_OVFCHK (a, b), "ut8-add 1");`

File: test/unit/test_ovf.c [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+		// ok
+	} else {
+		mu_assert_eq(a, b, "ut8-add 1");
+	}
+	b = 2;
+	if (UT8_ADD_OVFCHK(a, b)) {
```
Comment (ret2libc) [inline test/unit/test_ovf.c:13]: `mu_assert_true (UT8_ADD_OVFCHK (a, b), "ut8-add 2");`.

Comment (ret2libc) [reply test/unit/test_ovf.c:13]: and so on for the others.

File: libr/include/r_types_overflow.h (hunk 1 above)
Comment (trufae) [reply libr/include/r_types_overflow.h:9]: this if0 code is from the stackoverflow page you shared, it's there for reference, basically because not all conditions are done, and not all cases are handled in the suite. but i wanted to push the pr before falling asleep to get some feedback and maybe catch an issue here or there

File: libr/include/r_types_overflow.h (hunk 3 above)
Comment (trufae) [reply libr/include/r_types_overflow.h:42]: see above

File: libr/include/r_types_overflow.h (hunk 2 above)
Comment (trufae) [reply libr/include/r_types_overflow.h:15]: see above

File: test/unit/test_ovf.c (hunk 5 above)
Comment (trufae) [reply test/unit/test_ovf.c:7]: yeah i didnt liked this approach.ill change that now

File: libr/include/r_types_overflow.h (hunk 1 above)
Comment (ret2libc) [reply libr/include/r_types_overflow.h:9]: yeah it is ok to keep the code inside the `#if 0`. I'm just saying, just use comments instead of using `#if 0` to comment stuff.

File: libr/include/r_types_overflow.h (hunk 3 above)
Comment (ret2libc) [reply libr/include/r_types_overflow.h:42]: see above what? I didn't see above why this was placed at the end and not where more relevant.

**Comment by trufae** [comment]
chk again

**Comment by trufae** [comment]
But still not complete! Can you test those macros and review the tests? i dont think they are fine in every situation yet. also not sure if we want to handle underflow and overflow in the same check (imho its a good idea for simplicity)

**Comment by ret2libc** [comment]
> But still not complete! Can you test those macros and review the tests? i dont think they are fine in every situation yet. also not sure if we want to handle underflow and overflow in the same check (imho its a good idea for simplicity)

I will have a look.

I agree it's better to have underfolow/overflow in the same check.

File: libr/include/r_types_overflow.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
+#ifndef R_TYPES_OVERFLOW_H
+#define R_TYPES_OVERFLOW_H
+
+// ADD
+// if ((x > 0) && (a > INT_MAX - x)) /* `a + x` would overflow */;
+// if ((x < 0) && (a < INT_MIN - x)) /* `a + x` would underflow */;
+#define SZT_ADD_OVFCHK(x,y) ((SIZE_MAX - (x)) < (y))
+#define SSZT_ADD_OVFCHK(x,y) ((SSIZE_MAX - (x)) < (y))
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:8]: also here you need the `x > 0` part i think

File: test/unit/test_ovf.c [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	mu_assert_false (UT8_ADD_OVFCHK (250, 2), "ut8-add 2");
+	mu_assert_false (UT16_ADD_OVFCHK (ST16_MAX, 2), "ut16-add 2");
+	mu_assert_true (ST16_ADD_OVFCHK (ST16_MAX, 2), "st16-add 2");
+	mu_assert_true (ST16_ADD_OVFCHK (UT16_MAX, 1), "st16-add 2");
+	mu_assert_true (ST16_ADD_OVFCHK (1, ST16_MAX), "st16-add 3");
+	mu_end;
+}
+
+int test_underflow_sub(void) {
+	mu_assert_true (ST16_SUB_OVFCHK (10, 210), "st16-sub-underflow");
```
Comment (ret2libc) [inline test/unit/test_ovf.c:17]: shouldn't this be `false`?

10 - 210 = -200, which can be represented in ST16, no?

Comment (trufae) [reply test/unit/test_ovf.c:17]: right, this seems to be a bug

**Comment by trufae** [comment]
@ret2libc i catched a couple of more bugs and added more tests

File: test/unit/test_ovf.c (hunk 8 above)
Comment (ret2libc) [reply test/unit/test_ovf.c:24]: this isn't fixed yet, is it?

**Comment by trufae** [comment]
10-210 produces an underflow, so it asserts as true, lgtm, isnt?

File: test/unit/test_ovf.c (hunk 8 above)
Comment (ret2libc) [reply test/unit/test_ovf.c:24]: @trufae see https://github.com/radareorg/radare2/pull/17216#discussion_r454868061 . Am I missing anything?

Comment (trufae) [reply test/unit/test_ovf.c:24]: It is an underflow because it changes the sign

File: libr/include/r_types_overflow.h (hunk 4 above)
Comment (trufae) [reply libr/include/r_types_overflow.h:37]: They do now. It was wip

File: libr/include/r_types_overflow.h [hunk 9]
```diff
@@ -0,0 +1,53 @@
+#ifndef R_TYPES_OVERFLOW_H
+#define R_TYPES_OVERFLOW_H
+
+// TODO: Use CLANG/GCC builtins if available: __builtin_mul_overflow
+
+// ADD
+// if ((x > 0) && (a > INT_MAX - x)) /* `a + x` would overflow */;
+// if ((x < 0) && (a < INT_MIN - x)) /* `a + x` would underflow */;
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:8]: it seems like you are not considering the `INT_MIN` part in all the signed overflow check.
For example:
```C
#define SSZT_ADD_OVFCHK(a,x) ((((x) > 0) && ((a) > SSIZE_MAX - (x))) || (((x) < 0) && (a) < - (x)))
```
this should probably be,
```C
#define SSZT_ADD_OVFCHK(a,x) ((((x) > 0) && ((a) > SSIZE_MAX - (x))) || (((x) < 0) && (a) < SSIZE_MIN - (x)))
```
This is missed in many other checks and I think it's the reason for https://github.com/radareorg/radare2/pull/17216/files#diff-34737e40da2df87d09d8a4ffbbaf4602R24

Comment (trufae) [reply libr/include/r_types_overflow.h:8]: there's no SSIZE_MIN, but i can invent it

Comment (ret2libc) [reply libr/include/r_types_overflow.h:8]: if we don't have SSIZE_MIN this kind of check is not going to work :(

Comment (trufae) [reply libr/include/r_types_overflow.h:8]: done

File: libr/include/r_types_overflow.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+#define SZT_ADD_OVFCHK(x,y) ((SIZE_MAX - (x)) < (y))
+#define SSZT_ADD_OVFCHK(a,x) ((((x) > 0) && ((a) > SSIZE_MAX - (x))) || (((x) < 0) && (a) < SSIZE_MIN - (x)))
+#define UT64_ADD_OVFCHK(x,y) ((UT64_MAX - (x)) < (y))
+#define ST64_ADD_OVFCHK(a,x) ((((x) > 0) && ((a) > ST64_MAX - (x))) || (((x) < 0) && (a) < ST64_MIN - (x)))
+#define UT32_ADD_OVFCHK(x,y) ((UT32_MAX - (x)) < (y))
+#define ST32_ADD_OVFCHK(a,x) ((((x) > 0) && ((a) > ST32_MAX - (x))) || (((x) < 0) && (a) < ST32_MIN - (x)))
+#define UT16_ADD_OVFCHK(x,y) ((UT16_MAX - (x)) < (y))
+#define ST16_ADD_OVFCHK(a,b) ( \
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:16]: just for consistency put even this define on a single line. it should be as long as the other ones, more or less

File: libr/include/r_types_overflow.h [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+#define UT8_SUB_OVFCHK(a,b) UT8_ADD_OVFCHK(a,-(b))
+#define ST8_SUB_OVFCHK(a,b) ST8_ADD_OVFCHK(a,-(b))
+
+// MUL
+//if ((a == -1) && (x == INT_MIN)) /* `a * x` can overflow */
+//if ((x == -1) && (a == INT_MIN)) /* `a * x` (or `a / x`) can overflow */
+//if (a > INT_MAX / x) /* `a * x` would overflow */;
+//if ((a < INT_MIN / x)) /* `a * x` would underflow */;
+#define SZT_MUL_OVFCHK(x,y) ((y) && (x) > (SIZE_MAX / (y)))
+#define SSZT_MUL_OVFCHK(x,y) (((y) && (x) > (SSIZE_MAX / (y))) || ((((x) == -1) && ((y) == (SSIZE_MAX + 1)))) || ((y) == -1 && (x) == (SSIZE_MAX + 1)))
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:43]: use SSIZE_MIN instead of SSIZE_MAX since we now have it.

File: test/unit/test_ovf.c [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+	mu_assert_true (UT16_SUB_OVFCHK (10, 11), "st16-add 10-11");
+	mu_end;
+}
+
+int test_overflow_mul(void) {
+	mu_assert_true (UT8_MUL_OVFCHK (16, 32), "ut8-mul");
+	mu_assert_false (UT8_MUL_OVFCHK (16, 2), "ut8-mul 2");
+	mu_assert_true (ST8_MUL_OVFCHK (16, 100), "st8-mul 3");
+	mu_assert_false (ST8_MUL_OVFCHK (16, 1), "st8-mul 4");
+	mu_assert_true (ST8_MUL_OVFCHK (-1, -2), "ut8-mul sign overflow");
```
Comment (ret2libc) [inline test/unit/test_ovf.c:45]: I think this should not overflow.
`-1 * -2 == 2`, which can be represented with 8 bits.

File: test/unit/test_ovf.c [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+int test_underflow_sub(void) {
+	mu_assert_false (ST16_SUB_OVFCHK (10, 210), "st16-sub-sign-underflow");
+	mu_assert_true (UT16_SUB_OVFCHK (10, 210), "ut16-sub-underflow");
+	mu_assert_true (ST16_SUB_OVFCHK (ST16_MIN, 210), "st16-sub-underflow");
+	mu_end;
+}
+
+int test_underflow_add(void) {
+	mu_assert_false (ST16_SUB_OVFCHK (10, -210), "st16-add");
```
Comment (ret2libc) [inline test/unit/test_ovf.c:32]: change comments `s/-add/-sub/g`

File: libr/include/r_types_overflow.h [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+#define ST32_SUB_OVFCHK(a,b) ST32_ADD_OVFCHK(a,-(b))
+#define UT16_SUB_OVFCHK(a,b) UT16_ADD_OVFCHK(a,-(b))
+#define ST16_SUB_OVFCHK(a,b) ST16_ADD_OVFCHK(a,-(b))
+#define UT8_SUB_OVFCHK(a,b) UT8_ADD_OVFCHK(a,-(b))
+#define ST8_SUB_OVFCHK(a,b) ST8_ADD_OVFCHK(a,-(b))
+
+// MUL
+#define UNSIGNED_MUL_OVERFLOW_CHECK(overflow_name, type_base, type_min, type_max) \
+static inline bool overflow_name(type_base a, type_base b) { \
+	return (a > 0 && b > 0) ? (a > type_max / b): 0; \
```
[the remark this answers is not in the document]
Comment (trufae) [reply libr/include/r_types_overflow.h:37]: false is not yet there, thats why i used 0, but i solved it in a cleaner way

File: libr/include/r_types_overflow.h [hunk 15]
```diff
@@ -34,7 +34,7 @@
 // MUL
 #define UNSIGNED_MUL_OVERFLOW_CHECK(overflow_name, type_base, type_min, type_max) \
 static inline bool overflow_name(type_base a, type_base b) { \
-	return (a > 0 && b > 0) ? (a > type_max / b): 0; \
+	return (a > 0 && b > 0 && a > type_max / b); \
```
Comment (ret2libc) [inline libr/include/r_types_overflow.h:37]: even better :D

**Comment by XVilka** [comment]
```
cl : Command line warning D9002: ignoring unknown option '-std=c99' [C:\projects\radare2\build\shlr\cafa177@@tree_sitter@sta.vcxproj]
c:\projects\radare2\shlr\tree-sitter\lib\src\alloc.h(48): warning C4477: 'fprintf' : format string '%lu' requires an argument of type 'unsigned long', but variadic argument 1 has type 'size_t' [C:\projects\radare2\build\shlr\cafa177@@tree_sitter@sta.vcxproj]
c:\projects\radare2\shlr\tree-sitter\lib\src\alloc.h(57): warning C4477: 'fprintf' : format string '%lu' requires an argument of type 'unsigned long', but variadic argument 1 has type 'size_t' [C:\projects\radare2\build\shlr\cafa177@@tree_sitter@sta.vcxproj]
c:\projects\radare2\shlr\tree-sitter\lib\src\alloc.h(66): warning C4477: 'fprintf' : format string '%lu' requires an argument of type 'unsigned long', but variadic argument 1 has type 'size_t' [C:\projects\radare2\build\shlr\cafa177@@tree_sitter@sta.vcxproj]
c:\projects\radare2\shlr\sdb\src\sdb.c(1169): warning C4090: 'function': different 'const' qualifiers [C:\projects\radare2\build\sdb@exe.vcxproj]
c:\projects\radare2\shlr\capstone\arch\aarch64\aarch64genasmwriter.inc(6015): warning C4244: 'function': conversion from 'double' to 'int64_t', possible loss of data [C:\projects\radare2\build\shlr\cafa177@@r2capstone@sta.vcxproj]
c:\projects\radare2\shlr\sdb\src\sdb.c(1169): warning C4090: 'function': different 'const' qualifiers [C:\projects\radare2\build\r2sdb@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2146: syntax error: missing ')' before identifier 'a' [C:\projects\radare2\build\shlr\cafa177@@r2lz4@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2061: syntax error: identifier 'a' [C:\projects\radare2\build\shlr\cafa177@@r2lz4@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ';' [C:\projects\radare2\build\shlr\cafa177@@r2lz4@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ',' [C:\projects\radare2\build\shlr\cafa177@@r2lz4@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ')' [C:\projects\radare2\build\shlr\cafa177@@r2lz4@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2146: syntax error: missing ')' before identifier 'a' [C:\projects\radare2\build\shlr\cafa177@@r2zip@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2061: syntax error: identifier 'a' [C:\projects\radare2\build\shlr\cafa177@@r2zip@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ';' [C:\projects\radare2\build\shlr\cafa177@@r2zip@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ',' [C:\projects\radare2\build\shlr\cafa177@@r2zip@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ')' [C:\projects\radare2\build\shlr\cafa177@@r2zip@sta.vcxproj]
c:\projects\radare2\shlr\zip\zip\config.h(207): error C2371: 'size_t': redefinition; different basic types [C:\projects\radare2\build\shlr\cafa177@@r2zip@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2146: syntax error: missing ')' before identifier 'a' [C:\projects\radare2\build\r2spp@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2061: syntax error: identifier 'a' [C:\projects\radare2\build\r2spp@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ';' [C:\projects\radare2\build\r2spp@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ',' [C:\projects\radare2\build\r2spp@sta.vcxproj]
c:\projects\radare2\libr\include\r_types_overflow.h(50): error C2059: syntax error: ')' [C:\projects\radare2\build\r2spp@sta.vcxproj]
    7 Warning(s)
    16 Error(s)
```
btw, @ret2libc maybe these first MSVC warnings should be reported to tree-sitter itself?

**Comment by ret2libc** [comment]
> btw, @ret2libc maybe these first MSVC warnings should be reported to tree-sitter itself?

Done. https://github.com/tree-sitter/tree-sitter/pull/696

**Comment by trufae** [comment]
removed that ssize-t

**Comment by trufae** [comment]
@ret2libc @GustavoLCR ok to merge now?
